### ISSUE TITLE
Unexpected behaviour while appending a python list inside ti.kernel

### ISSUE THREAD

**Issue by xchhuang** [taichi-dev/taichi#2271]
**Describe the bug**
I try the following code to append a python list inside a taichi kernel. But whatever ```N``` is, ```array_ti``` would append once.

**To Reproduce**
Here is my example:

```py
import taichi as ti

ti.init(arch=ti.cpu)

N = 0
array_py = []
array_ti = []


# expected behaviour
def test_append_py():
    if N > 0:
        array_py.append(1)


# weird behaviour
@ti.kernel
def test_append_ti():
    if N > 0:
        array_ti.append(ti.Vector([1, 1]))


# tests
print('array_py before:', array_py)
for i in range(10):
    test_append_py()
print('array_py after:', array_py)

print('array_ti before:', array_ti)
for i in range(10):
    test_append_ti()
print('array_ti after:', array_ti)

```

**Log/Screenshots**

```
[Taichi] mode=release
[Taichi] version 0.7.14, llvm 10.0.0, commit 58feee37, win, python 3.7.10
[I 04/17/21 11:36:27.725] [shell.py:_shell_pop_print@31] Graphical python shell detected, using wrapped sys.stdout
[Taichi] Starting on arch=x64
array_py before: []
array_py after: []
array_ti before: []
[Taichi] materializing...
array_ti after: [[<ti.Expr> <ti.Expr>]]

```

Is this behaviour expected? If yes, is there anyway to dynamically append a python list? Thanks in advance.

**Comment by k-ye** [comment]
Hi,

Thanks for reporting this. Please do not use python data structures inside `@ti.kernel`. The code within a `@ti.kernel` is offloaded to and executed by the Taichi's C++ runtime, and has nothing to do with the Python interpretor. (Ideally Taichi should throw an error in this scenario to show exactly what has gone wrong.)

**Comment by xchhuang** [comment]
Hi @k-ye ,
Thanks for your clarification. And is it right that we can **read** python data structure(e.g., a list of ```ti.field```) inside ```@ti.kernel```, but not write?

**Comment by k-ye** [comment]
>  And is it right that we can read python data structure(e.g., a list of ti.field) inside @ti.kernel, but not write?
No, unfortunately that is not how taichi works. As explained above, the taichi kernel is *not* executed within the Python context at all. Even if this example itself doesn't crash, it will not match what you'd normally expect.

When you call Taichi kernel, what happens internally is that taichi JIT **compiles** the Python source code following a standard compiler approach. That is, Taichi extracts the abstract syntax tree (AST) of the `@ti.kernel` source code , transforms it into a lower-level IR format, does various optimizations and finally emits backend-native instructions (PTX for CUDA, for example). What you've observed in `array_ti` after calling `test_append_ti()` is an accidental effect during the AST transformation (internally, the transformed kernel is run once to generate the lower-level IR). If handled more robustly, this should throw out an error saying that Python scope data structures/functions are used.

TLDR; Please stick with https://taichi.readthedocs.io/en/latest/syntax.html and do not mix python-scoped values with taichi kernels.

**Comment by xchhuang** [comment]
I see, thank you!
